### PR TITLE
Added Shake128/256 Digest support

### PR DESCRIPTION
**Pull request by damon-kwok**
Added Shake128/256 Digest and testcase.

### CODE REVIEW

**Comment by damon-kwok** [comment]
```sh
Run make test config=debug ssl=0.9.0
mkdir -p build/debug
corral fetch
fetch: fetching from /__w/crypto/crypto
corral run -- ponyc --debug -Dopenssl_0.9.0 -o build/debug crypto
run: ponyc --debug -Dopenssl_0.9.0 -o build/debug crypto
  exit: 255
  out: 
  err: Building builtin -> /root/.local/share/ponyup/ponyc-release-0.35.0-x86_64-linux-musl/packages/builtin
Building crypto -> /__w/crypto/crypto/crypto
Building format -> /root/.local/share/ponyup/ponyc-release-0.35.0-x86_64-linux-musl/packages/format
Building collections -> /root/.local/share/ponyup/ponyc-release-0.35.0-x86_64-linux-musl/packages/collections
Building ponytest -> /root/.local/share/ponyup/ponyc-release-0.35.0-x86_64-linux-musl/packages/ponytest
Building time -> /root/.local/share/ponyup/ponyc-release-0.35.0-x86_64-linux-musl/packages/time
Building random -> /root/.local/share/ponyup/ponyc-release-0.35.0-x86_64-linux-musl/packages/random
Generating
 Reachability
 Selector painting
 Data prototypes
 Data types
 Function prototypes
 Functions
 Descriptors
Writing build/debug/crypto.o
Linking build/debug/crypto
/__w/crypto/crypto/crypto/digest.pony:110: error: undefined reference to 'EVP_shake128'
/__w/crypto/crypto/crypto/digest.pony:122: error: undefined reference to 'EVP_shake256'
clang-9: error: linker command failed with exit code 1 (use -v to see invocation)

```
`ssl 0.9.0` is too old,  it does not support `EVP_shake128` and `EVP_shake256`.
`ssl 3.0` will be released soon, we should probably limit the minimum version to 1.1.x
See:
https://wiki.openssl.org/index.php/OpenSSL_3.0
https://wiki.openssl.org/index.php/OpenSSL_3.0

File: crypto/digest.pony [hunk 1]
```diff
@@ -97,6 +97,30 @@ class Digest
...
+    """
+    Use the Shake128 algorithm to calculate the hash.
+    """
+    _digest_size = 16
+    ifdef "openssl_1.1.x" then
+      _ctx = @EVP_MD_CTX_new[Pointer[_EVPCTX]]()
+    else
+      _ctx = @EVP_MD_CTX_create[Pointer[_EVPCTX]]()
+    end
+    @EVP_DigestInit_ex[None](_ctx, @EVP_shake128[Pointer[_EVPMD]](), USize(0))
```
Comment (mfelsche) [inline crypto/digest.pony:110]: As this algorithm is not supported in openssl 0.9.0 you can move this inside the `ifdef "openssl_1.1.x"` and add a `compile_error` in the else clause.

File: crypto/_test.pony [hunk 2]
```diff
@@ -122,3 +122,17 @@ class iso _TestDigest is UnitTest
       "ee26b0dd4af7e749aa1a8ee3c10ae9923f618980772e473f8819a5d4940e0db2" +
       "7ac185f8a0e1d5f84f88bc887fd67b143732c304cc5fa9ad8e6f57f50028a8ff",
       ToHexString(d'.final()))
+
+    let s = Digest.shake128()
```
Comment (mfelsche) [inline crypto/_test.pony:126]: you can put an `ifdef "openssl_1.1.x"` around those two tests to guard them against old openssl versions.

Comment (mfelsche) [reply crypto/_test.pony:126]: Also putting the tests for different algorithms in their own test class would make the test setup much cleaner. But as it was like this already, no need to do it in this PR.

Comment (damon-kwok) [reply crypto/_test.pony:126]: NP

**Comment by SeanTAllen** [comment]
@mfelsche you merged this without adding a changelog label. please add a CHANGELOG entry manually.

**Comment by SeanTAllen** [comment]
I believe libressl supports shake128 but also supports the openssl 0.9.0 API. I think this PR isn't correct in its current form.

We should add additional PR tests openssl 1.1, openssl 0.9, and libressl ASAP. Nothing that says something is a compile error should be merged before that is done.

Given how libressl and openssl 0.9 have diverged, they might need their own compile time constants.

**Comment by SeanTAllen** [comment]
We don't have 1.1 testing at all for PRs so there's no guarantee that this works as the test wouldnt have run.

**Comment by SeanTAllen** [comment]
I've opened an issue to add the additional CI testing. 

https://github.com/ponylang/crypto/issues/29

until that is closed, we should be merging anything that has `compile error` in it. And probably shouldn't merge anything else until that is addressed in general.

**Comment by SeanTAllen** [comment]
It appears that libressl doesnt support Shake128.

**Comment by SeanTAllen** [comment]
This should have been squashed on merging. All the commits should have been 1 commit.

**Comment by SeanTAllen** [comment]
This is missing non-digest tests for both Shake128 and Shake256

**Comment by SeanTAllen** [comment]
Sorry I was wrong, the non-digest tests were added as part of the digest. Its still missing tests.

**Pull request merged**